// @firecms/neat/dist/index.es.js
const ht = `void main() {
vUv = uv;
vPosition = position;
float waveOffset = -u_y_offset * u_y_offset_wave_multiplier;
float colorOffset = -u_y_offset * u_y_offset_color_multiplier;
float flowOffset = -u_y_offset * u_y_offset_flow_multiplier;
v_displacement_amount = cnoise( vec3(
u_wave_frequency_x * position.x + u_time,
u_wave_frequency_y * (position.y + waveOffset) + u_time,
u_time
));
vec2 baseUv = vUv;
baseUv.y += flowOffset / u_plane_height;
vec2 flowUv = baseUv;
if (u_flow_enabled > 0.5) {
if (u_flow_ease > 0.0 || u_flow_distortion_a > 0.0) {
vec2 ppp = -1.0 + 2.0 * baseUv;
ppp += 0.1 * cos((1.5 * u_flow_scale) * ppp.yx + 1.1 * u_time + vec2(0.1, 1.1));
ppp += 0.1 * cos((2.3 * u_flow_scale) * ppp.yx + 1.3 * u_time + vec2(3.2, 3.4));
ppp += 0.1 * cos((2.2 * u_flow_scale) * ppp.yx + 1.7 * u_time + vec2(1.8, 5.2));
ppp += u_flow_distortion_a * cos((u_flow_distortion_b * u_flow_scale) * ppp.yx + 1.4 * u_time + vec2(6.3, 3.9));
float r = length(ppp);
flowUv = mix(baseUv, vec2(baseUv.x * (1.0 - u_flow_ease) + r * u_flow_ease, baseUv.y), u_flow_ease);
}
}
vFlowUv = flowUv;
vec3 color = u_colors[0].color;
vec3 distortedPos = position;
if (u_flat_shading < 0.5) {
if (u_flow_enabled > 0.5) {
if (u_flow_ease > 0.0 || u_flow_distortion_a > 0.0) {
vec3 ppp = position / 25.0;
ppp.xyz += 0.1 * cos((1.5 * u_flow_scale) * ppp.yxz + 1.1 * u_time + vec3(0.1, 1.1, 2.1));
ppp.xyz += 0.1 * cos((2.3 * u_flow_scale) * ppp.zxy + 1.3 * u_time + vec3(3.2, 3.4, 1.2));
ppp.xyz += 0.1 * cos((2.2 * u_flow_scale) * ppp.yxz + 1.7 * u_time + vec3(1.8, 5.2, 3.1));
ppp.xyz += u_flow_distortion_a * cos((u_flow_distortion_b * u_flow_scale) * ppp.zxy + 1.4 * u_time + vec3(6.3, 3.9, 4.5));
float r = length(ppp);
distortedPos = mix(position, vec3(
position.x * (1.0 - u_flow_ease) + r * u_flow_ease * 25.0,
position.y,
position.z * (1.0 - u_flow_ease) + r * u_flow_ease * 25.0
), u_flow_ease);
}
}
}
vec3 noise_cord;
if (u_flat_shading < 0.5) {
noise_cord = vec3(distortedPos.x / 50.0, (distortedPos.y + colorOffset) / 50.0, distortedPos.z / 50.0);
} else {
vec2 adjustedUv = flowUv;
adjustedUv.y += colorOffset / u_plane_height;
noise_cord = vec3(adjustedUv, 0.0);
}
const float minNoise = .0;
const float maxNoise = .9;
for (int i = 1; i < 6; i++) {
if (i < u_colors_count) {
if (u_colors[i].is_active > 0.5) {
float noiseFlow = (1. + float(i)) / 30.;
float noiseSpeed = (1. + float(i)) * 0.11;
float noiseSeed = 13. + float(i) * 7.;
float noise_z = u_time * noiseSpeed;
if (u_flat_shading < 0.5) {
noise_z = noise_cord.z * u_color_pressure.x * u_color_pressure.x + u_time * noiseSpeed;
}
float noise = snoise(
vec3(
noise_cord.x * u_color_pressure.x * u_color_pressure.x + u_time * noiseFlow * 2.,
noise_cord.y * u_color_pressure.y * u_color_pressure.y,
noise_z
) + noiseSeed
) - (.1 * float(i)) + (.5 * u_color_blending);
noise = clamp(noise, minNoise, maxNoise + float(i) * 0.02);
color = mix(color, u_colors[i].color, smoothstep(0.0, u_color_blending, noise));
}
}
}
v_color = color;
vec3 newPosition = position + normal * v_displacement_amount * u_wave_amplitude;
vec4 mvPosition = modelViewMatrix * vec4(newPosition, 1.0);
vViewPosition = mvPosition.xyz;
vNormal = normalize((modelViewMatrix * vec4(normal, 0.0)).xyz);
gl_Position = projectionMatrix * mvPosition;
v_new_position = gl_Position;
}`, dt = `float random(vec2 p) {
return fract(sin(dot(p, vec2(12.9898,78.233))) * 43758.5453);
}
float fbm(vec3 x) {
float value = 0.0;
float amplitude = 0.5;
float frequency = 1.0;
for (int i = 0; i < 2; i++) {
value += amplitude * snoise(x * frequency);
frequency *= 2.0;
amplitude *= 0.5;
}
return value;
}
vec3 hsl2rgb(float h, float s, float l) {
vec3 rgb = clamp(abs(mod(h * 6.0 + vec3(0.0, 4.0, 2.0), 6.0) - 3.0) - 1.0, 0.0, 1.0);
return l + s * (rgb - 0.5) * (1.0 - abs(2.0 * l - 1.0));
}
void main() {
vec2 finalUv = vFlowUv;
vec3 baseColor;
float texAlpha = 1.0;
if (u_enable_procedural_texture > 0.5) {
if (u_flat_shading < 0.5) {
float parallaxFactor = 0.25;
float scrollOffset = (u_y_offset * u_y_offset_color_multiplier) * parallaxFactor;
vec3 scrolledPos = vPosition;
scrolledPos.y -= scrollOffset;
vec3 p = (scrolledPos * 1.5) / 50.0;
vec2 uvX = p.yz + vec2(0.5);
vec2 uvY = p.zx + vec2(0.5);
vec2 uvZ = p.xy + vec2(0.5);
vec4 colX = texture2D(u_procedural_texture, uvX);
vec4 colY = texture2D(u_procedural_texture, uvY);
vec4 colZ = texture2D(u_procedural_texture, uvZ);
vec3 n = normalize(vNormal);
vec3 blendWeights = abs(n);
blendWeights = blendWeights / (blendWeights.x + blendWeights.y + blendWeights.z + 0.0001);
vec4 texSample = colX * blendWeights.x + colY * blendWeights.y + colZ * blendWeights.z;
baseColor = texSample.rgb;
if (u_transparent_texture_void > 0.5) {
texAlpha = texSample.a;
}
} else {
vec2 ppp = -1.0 + 2.0 * finalUv;
ppp += 0.1 * cos((1.5 * u_flow_scale) * ppp.yx + 1.1 * u_time + vec2(0.1, 1.1));
ppp += 0.1 * cos((2.3 * u_flow_scale) * ppp.yx + 1.3 * u_time + vec2(3.2, 3.4));
ppp += 0.1 * cos((2.2 * u_flow_scale) * ppp.yx + 1.7 * u_time + vec2(1.8, 5.2));
ppp += u_flow_distortion_a * cos((u_flow_distortion_b * u_flow_scale) * ppp.yx + 1.4 * u_time + vec2(6.3, 3.9));
float r = length(ppp);
float vx = (finalUv.x * u_texture_ease) + (r * (1.0 - u_texture_ease));
float vy = (finalUv.y * u_texture_ease) + (0.0 * (1.0 - u_texture_ease));
vec2 texUv = vec2(vx, vy);
float parallaxFactor = 0.25;
texUv.y -= (u_y_offset * u_y_offset_color_multiplier / u_plane_height) * parallaxFactor;
texUv *= 1.5;
vec4 texSample = texture2D(u_procedural_texture, texUv);
baseColor = texSample.rgb;
if (u_transparent_texture_void > 0.5) {
texAlpha = texSample.a;
}
}
} else {
baseColor = v_color;
}
vec3 color = baseColor;
if (u_domain_warp_enabled > 0.5) {
vec3 p;
if (u_flat_shading < 0.5) {
p = vec3((vPosition / 50.0 + vec3(0.5)) * u_domain_warp_scale);
p.z += u_time * 0.15;
} else {
p = vec3(finalUv * u_domain_warp_scale, u_time * 0.15);
}
vec2 q = vec2(fbm(p), fbm(p + vec3(5.2, 1.3, 0.0)));
float f = fbm(p + vec3(4.0 * q, 0.0));
vec3 warpColor = color * (1.0 + f * 0.8 * u_domain_warp_intensity);
float pattern = clamp(f * f * f + 0.6 * f * f + 0.5 * f, 0.0, 1.0);
color = mix(color, warpColor * (0.6 + pattern * 0.8), u_domain_warp_intensity * 0.7);
}
vec3 normal = normalize(vNormal);
vec3 viewDir = vec3(0.0, 0.0, 1.0);
float ndotv = dot(normal, viewDir);
if (u_shape_type > 0.5 && u_shape_type < 3.5) {
if (ndotv < 0.0) {
discard;
}
} else {
if (ndotv < 0.0) {
normal = -normal;
ndotv = -ndotv;
}
}
vec3 lightDir = normalize(vec3(1.0, 1.0, 1.0));
float diffuse = max(dot(normal, lightDir), 0.0);
vec3 halfDir = normalize(lightDir + viewDir);
float specular = pow(max(dot(normal, halfDir), 0.0), 32.0);
if (u_flat_shading > 0.5) {
color += v_displacement_amount * u_highlights;
float heightShadow = 1.0 - v_displacement_amount;
color -= heightShadow * heightShadow * u_shadows;
} else {
color += specular * u_highlights;
color += v_displacement_amount * u_highlights * 0.5;
float heightShadow = 1.0 - v_displacement_amount;
color -= heightShadow * heightShadow * u_shadows * 0.5;
color -= (1.0 - diffuse) * u_shadows * 0.5;
}
color = saturation(color, 1.0 + u_saturation);
color = color * u_brightness;
if (u_iridescence_enabled > 0.5) {
float hue = fract(v_displacement_amount * 0.5 + 0.5 + u_time * u_iridescence_speed * 0.05);
vec3 iriColor = hsl2rgb(hue, 0.8, 0.6);
color = mix(color, iriColor, u_iridescence_intensity * abs(v_displacement_amount) * 0.6);
}
if (u_fresnel_enabled > 0.5) {
float slope = 1.0 - abs(v_displacement_amount);
float fresnel = pow(max(slope, 0.0), u_fresnel_power);
color += u_fresnel_color * fresnel * u_fresnel_intensity;
}
if (u_vignette_intensity > 0.0) {
vec2 vigUv = vUv;
if (u_flat_shading < 0.5) {
vigUv = (v_new_position.xy / v_new_position.w) * 0.5 + vec2(0.5);
}
float dist = length(vigUv - vec2(0.5));
float vig = smoothstep(u_vignette_radius, u_vignette_radius * 0.3, dist);
color *= mix(1.0, vig, u_vignette_intensity);
}
if (u_bloom_intensity > 0.0) {
float luma = dot(color, vec3(0.2126, 0.7152, 0.0722));
float bloomMask = smoothstep(u_bloom_threshold, 1.0, luma);
color += color * bloomMask * u_bloom_intensity;
}
if (u_chromatic_aberration > 0.0) {
float caAmount = u_chromatic_aberration * 0.008;
vec2 caUv = vUv;
if (u_flat_shading < 0.5) {
caUv = (v_new_position.xy / v_new_position.w) * 0.5 + vec2(0.5);
}
float dist = length(caUv - vec2(0.5));
float rShift = v_displacement_amount + caAmount * dist;
float bShift = v_displacement_amount - caAmount * dist;
color.r *= 1.0 + rShift * caAmount * 10.0;
color.b *= 1.0 - bShift * caAmount * 10.0;
}
float grain = 0.0;
if (u_grain_intensity > 0.0) {
vec2 noiseCoords = gl_FragCoord.xy / u_grain_scale;
if (u_grain_speed != 0.0 || u_flat_shading > 0.5) {
grain = fbm(vec3(noiseCoords, u_time * u_grain_speed));
} else {
grain = random(noiseCoords) - 0.5;
}
grain = grain * 0.5 + 0.5;
grain -= 0.5;
grain = (grain > u_grain_sparsity) ? grain : 0.0;
grain *= u_grain_intensity;
}
color += vec3(grain);
float edgeAlpha = 1.0;
if (u_silhouette_fade > 0.0 && u_flat_shading < 0.5) {
edgeAlpha = smoothstep(0.0, u_silhouette_fade, ndotv);
}
if (u_shape_type == 3.0) {
float vFade = smoothstep(0.0, u_cylinder_fade, vUv.y) * smoothstep(1.0, 1.0 - u_cylinder_fade, vUv.y);
edgeAlpha *= vFade;
} else if (u_shape_type == 4.0) {
float uFade = smoothstep(0.0, u_ribbon_fade, vUv.x) * smoothstep(1.0, 1.0 - u_ribbon_fade, vUv.x);
float vFade = smoothstep(0.0, u_ribbon_fade, vUv.y) * smoothstep(1.0, 1.0 - u_ribbon_fade, vUv.y);
edgeAlpha *= uFade * vFade;
}
edgeAlpha *= texAlpha;
gl_FragColor = vec4(color, edgeAlpha);
}`;
function mt() {
  return `precision highp float;
attribute vec3 position;
attribute vec3 normal;
attribute vec2 uv;
uniform mat4 modelViewMatrix;
uniform mat4 projectionMatrix;
varying vec2 vUv;
varying vec2 vFlowUv;
varying vec4 v_new_position;
varying vec3 v_color;
varying float v_displacement_amount;
varying vec3 vViewPosition;
varying vec3 vNormal;
varying vec3 vPosition;
uniform float u_time;
uniform vec2 u_resolution;
uniform vec2 u_color_pressure;
uniform float u_wave_frequency_x;
uniform float u_wave_frequency_y;
uniform float u_wave_amplitude;
uniform float u_plane_width;
uniform float u_plane_height;
uniform float u_color_blending;
uniform int u_colors_count;
struct ColorStop {
float is_active;
vec3 color;
float influence;
};
uniform ColorStop u_colors[6];
uniform float u_y_offset;
uniform float u_y_offset_wave_multiplier;
uniform float u_y_offset_color_multiplier;
uniform float u_y_offset_flow_multiplier;
uniform float u_flow_distortion_a;
uniform float u_flow_distortion_b;
uniform float u_flow_scale;
uniform float u_flow_ease;
uniform float u_flow_enabled;
uniform float u_fresnel_enabled;
uniform float u_fresnel_power;
uniform float u_fresnel_intensity;
uniform vec3 u_fresnel_color;
uniform float u_shape_type;
uniform float u_flat_shading;`;
}
function pt() {
  return `precision highp float;
varying vec2 vUv;
varying vec2 vFlowUv;
varying vec4 v_new_position;
varying vec3 v_color;
varying float v_displacement_amount;
varying vec3 vViewPosition;
varying vec3 vNormal;
varying vec3 vPosition;
uniform float u_time;
uniform vec2 u_resolution;
uniform float u_plane_height;
uniform float u_shadows;
uniform float u_highlights;
uniform float u_saturation;
uniform float u_brightness;
uniform float u_grain_intensity;
uniform float u_grain_sparsity;
uniform float u_grain_scale;
uniform float u_grain_speed;
uniform float u_y_offset;
uniform float u_y_offset_color_multiplier;
uniform float u_flow_distortion_a;
uniform float u_flow_distortion_b;
uniform float u_flow_scale;
uniform sampler2D u_procedural_texture;
uniform float u_enable_procedural_texture;
uniform float u_texture_ease;
uniform float u_domain_warp_enabled;
uniform float u_domain_warp_intensity;
uniform float u_domain_warp_scale;
uniform float u_vignette_intensity;
uniform float u_vignette_radius;
uniform float u_fresnel_enabled;
uniform float u_fresnel_power;
uniform float u_fresnel_intensity;
uniform vec3 u_fresnel_color;
uniform float u_iridescence_enabled;
uniform float u_iridescence_intensity;
uniform float u_iridescence_speed;
uniform float u_bloom_intensity;
uniform float u_bloom_threshold;
uniform float u_chromatic_aberration;
uniform float u_shape_type;
uniform float u_transparent_texture_void;
uniform float u_silhouette_fade;
uniform float u_cylinder_fade;
uniform float u_ribbon_fade;
uniform float u_flat_shading;`;
}
function re() {
  return `vec4 permute(vec4 x) {
return floor(fract(sin(x) * 43758.5453123) * 289.0);
}
vec4 taylorInvSqrt(vec4 r) {
return 1.79284291400159 - 0.85373472095314 * r;
}
vec3 fade(vec3 t) {
return t*t*t*(t*(t*6.0-15.0)+10.0);
}
float snoise(vec3 v) {
const vec2 C = vec2(1.0/6.0, 1.0/3.0) ;
const vec4 D = vec4(0.0, 0.5, 1.0, 2.0);
vec3 i = floor(v + dot(v, C.yyy) );
vec3 x0 = v - i + dot(i, C.xxx) ;
vec3 g = step(x0.yzx, x0.xyz);
vec3 l = 1.0 - g;
vec3 i1 = min( g.xyz, l.zxy );
vec3 i2 = max( g.xyz, l.zxy );
vec3 x1 = x0 - i1 + C.xxx;
vec3 x2 = x0 - i2 + C.yyy;
vec3 x3 = x0 - D.yyy;
vec4 p = permute( permute( permute(
i.z + vec4(0.0, i1.z, i2.z, 1.0 ))
+ i.y + vec4(0.0, i1.y, i2.y, 1.0 ))
+ i.x + vec4(0.0, i1.x, i2.x, 1.0 ));
float n_ = 0.142857142857;
vec3 ns = n_ * D.wyz - D.xzx;
vec4 j = p - 49.0 * floor(p * ns.z * ns.z);
vec4 x_ = floor(j * ns.z);
vec4 y_ = floor(j - 7.0 * x_ );
vec4 x = x_ *ns.x + ns.yyyy;
vec4 y = y_ *ns.x + ns.yyyy;
vec4 h = 1.0 - abs(x) - abs(y);
vec4 b0 = vec4( x.xy, y.xy );
vec4 b1 = vec4( x.zw, y.zw );
vec4 s0 = floor(b0)*2.0 + 1.0;
vec4 s1 = floor(b1)*2.0 + 1.0;
vec4 sh = -step(h, vec4(0.0));
vec4 a0 = b0.xzyw + s0.xzyw*sh.xxyy ;
vec4 a1 = b1.xzyw + s1.xzyw*sh.zzww ;
vec3 p0 = vec3(a0.xy,h.x);
vec3 p1 = vec3(a0.zw,h.y);
vec3 p2 = vec3(a1.xy,h.z);
vec3 p3 = vec3(a1.zw,h.w);
vec4 norm = taylorInvSqrt(vec4(dot(p0,p0), dot(p1,p1), dot(p2, p2), dot(p3,p3)));
p0 *= norm.x;
p1 *= norm.y;
p2 *= norm.z;
p3 *= norm.w;
vec4 m = max(0.6 - vec4(dot(x0,x0), dot(x1,x1), dot(x2,x2), dot(x3,x3)), 0.0);
m = m * m;
return 42.0 * dot( m*m, vec4( dot(p0,x0), dot(p1,x1),
dot(p2,x2), dot(p3,x3) ) );
}
float cnoise(vec3 P)
{
vec3 Pi0 = floor(P);
vec3 Pi1 = Pi0 + vec3(1.0);
vec3 Pf0 = fract(P);
vec3 Pf1 = Pf0 - vec3(1.0);
vec4 ix = vec4(Pi0.x, Pi1.x, Pi0.x, Pi1.x);
vec4 iy = vec4(Pi0.yy, Pi1.yy);
vec4 iz0 = Pi0.zzzz;
vec4 iz1 = Pi1.zzzz;
vec4 ixy = permute(permute(ix) + iy);
vec4 ixy0 = permute(ixy + iz0);
vec4 ixy1 = permute(ixy + iz1);
vec4 gx0 = ixy0 * (1.0 / 7.0);
vec4 gy0 = fract(floor(gx0) * (1.0 / 7.0)) - 0.5;
gx0 = fract(gx0);
vec4 gz0 = vec4(0.5) - abs(gx0) - abs(gy0);
vec4 sz0 = step(gz0, vec4(0.0));
gx0 -= sz0 * (step(0.0, gx0) - 0.5);
gy0 -= sz0 * (step(0.0, gy0) - 0.5);
vec4 gx1 = ixy1 * (1.0 / 7.0);
vec4 gy1 = fract(floor(gx1) * (1.0 / 7.0)) - 0.5;
gx1 = fract(gx1);
vec4 gz1 = vec4(0.5) - abs(gx1) - abs(gy1);
vec4 sz1 = step(gz1, vec4(0.0));
gx1 -= sz1 * (step(0.0, gx1) - 0.5);
gy1 -= sz1 * (step(0.0, gy1) - 0.5);
vec3 g000 = vec3(gx0.x,gy0.x,gz0.x);
vec3 g100 = vec3(gx0.y,gy0.y,gz0.y);
vec3 g010 = vec3(gx0.z,gy0.z,gz0.z);
vec3 g110 = vec3(gx0.w,gy0.w,gz0.w);
vec3 g001 = vec3(gx1.x,gy1.x,gz1.x);
vec3 g101 = vec3(gx1.y,gy1.y,gz1.y);
vec3 g011 = vec3(gx1.z,gy1.z,gz1.z);
vec3 g111 = vec3(gx1.w,gy1.w,gz1.w);
vec4 norm0 = taylorInvSqrt(vec4(dot(g000, g000), dot(g010, g010), dot(g100, g100), dot(g110, g110)));
g000 *= norm0.x;
g010 *= norm0.y;
g100 *= norm0.z;
g110 *= norm0.w;
vec4 norm1 = taylorInvSqrt(vec4(dot(g001, g001), dot(g011, g011), dot(g101, g101), dot(g111, g111)));
g001 *= norm1.x;
g011 *= norm1.y;
g101 *= norm1.z;
g111 *= norm1.w;
float n000 = dot(g000, Pf0);
float n100 = dot(g100, vec3(Pf1.x, Pf0.yz));
float n010 = dot(g010, vec3(Pf0.x, Pf1.y, Pf0.z));
float n110 = dot(g110, vec3(Pf1.xy, Pf0.z));
float n001 = dot(g001, vec3(Pf0.xy, Pf1.z));
float n101 = dot(g101, vec3(Pf1.x, Pf0.y, Pf1.z));
float n011 = dot(g011, vec3(Pf0.x, Pf1.yz));
float n111 = dot(g111, Pf1);
vec3 fade_xyz = fade(Pf0);
vec4 n_z = mix(vec4(n000, n100, n010, n110), vec4(n001, n101, n011, n111), fade_xyz.z);
vec2 n_yz = mix(n_z.xy, n_z.zw, fade_xyz.y);
float n_xyz = mix(n_yz.x, n_yz.y, fade_xyz.x);
return 2.2 * n_xyz;
}`;
}
function se() {
  return `vec3 saturation(vec3 rgb, float adjustment) {
const vec3 W = vec3(0.2125, 0.7154, 0.0721);
vec3 intensity = vec3(dot(rgb, W));
return mix(intensity, rgb, adjustment);
}`;
}
class fe {
  elements;
  constructor() {
    this.elements = new Float32Array([
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1
    ]);
  }
  identity() {
    const e = this.elements;
    return e[0] = 1, e[1] = 0, e[2] = 0, e[3] = 0, e[4] = 0, e[5] = 1, e[6] = 0, e[7] = 0, e[8] = 0, e[9] = 0, e[10] = 1, e[11] = 0, e[12] = 0, e[13] = 0, e[14] = 0, e[15] = 1, this;
  }
  translate(e, i, s) {
    return this.elements[12] += this.elements[0] * e + this.elements[4] * i + this.elements[8] * s, this.elements[13] += this.elements[1] * e + this.elements[5] * i + this.elements[9] * s, this.elements[14] += this.elements[2] * e + this.elements[6] * i + this.elements[10] * s, this.elements[15] += this.elements[3] * e + this.elements[7] * i + this.elements[11] * s, this;
  }
  rotateX(e) {
    const i = Math.cos(e), s = Math.sin(e), r = this.elements[4], t = this.elements[5], a = this.elements[6], o = this.elements[7], _ = this.elements[8], m = this.elements[9], w = this.elements[10], v = this.elements[11];
    return this.elements[4] = i * r + s * _, this.elements[5] = i * t + s * m, this.elements[6] = i * a + s * w, this.elements[7] = i * o + s * v, this.elements[8] = i * _ - s * r, this.elements[9] = i * m - s * t, this.elements[10] = i * w - s * a, this.elements[11] = i * v - s * o, this;
  }
  rotateY(e) {
    const i = Math.cos(e), s = Math.sin(e), r = this.elements[0], t = this.elements[1], a = this.elements[2], o = this.elements[3], _ = this.elements[8], m = this.elements[9], w = this.elements[10], v = this.elements[11];
    return this.elements[0] = i * r - s * _, this.elements[1] = i * t - s * m, this.elements[2] = i * a - s * w, this.elements[3] = i * o - s * v, this.elements[8] = s * r + i * _, this.elements[9] = s * t + i * m, this.elements[10] = s * a + i * w, this.elements[11] = s * o + i * v, this;
  }
  rotateZ(e) {
    const i = Math.cos(e), s = Math.sin(e), r = this.elements[0], t = this.elements[1], a = this.elements[2], o = this.elements[3], _ = this.elements[4], m = this.elements[5], w = this.elements[6], v = this.elements[7];
    return this.elements[0] = i * r + s * _, this.elements[1] = i * t + s * m, this.elements[2] = i * a + s * w, this.elements[3] = i * o + s * v, this.elements[4] = -s * r + i * _, this.elements[5] = -s * t + i * m, this.elements[6] = -s * a + i * w, this.elements[7] = -s * o + i * v, this;
  }
}
class vt {
  left;
  right;
  top;
  bottom;
  near;
  far;
  position;
  projectionMatrix;
  zoom;
  constructor(e, i, s, r, t, a) {
    this.left = e, this.right = i, this.top = s, this.bottom = r, this.near = t, this.far = a, this.position = [0, 0, 0], this.zoom = 1, this.projectionMatrix = new fe(), this.updateProjectionMatrix();
  }
  updateProjectionMatrix() {
    const e = 1 / (this.right - this.left), i = 1 / (this.top - this.bottom), s = 1 / (this.far - this.near), r = (this.right + this.left) * e, t = (this.top + this.bottom) * i, a = (this.far + this.near) * s;
    this.projectionMatrix.elements = new Float32Array([
      2 * e,
      0,
      0,
      0,
      0,
      2 * i,
      0,
      0,
      0,
      0,
      -2 * s,
      0,
      -r,
      -t,
      -a,
      1
    ]);
  }
}
function j(h, e, i, s = 50, r = 50, t = "plane", a = 1) {
  h.zoom = a;
  const o = e / i;
  if (t === "plane") {
    const w = e * i / 1e6 * s * r / 1.5, v = Math.sqrt(w * o), u = w / v;
    let x = -s / 2, l = Math.min((x + v) / 1.5, s / 2), f = r / 4, R = Math.max((f - u) / 2, -r / 4);
    if (o < 1) {
      const d = o;
      x = x * d, l = l * d;
      const b = 1.05;
      x = x * b, l = l * b, f = f * b, R = R * b;
    }
    h.left = x, h.right = l, h.top = f, h.bottom = R;
  } else {
    let _ = 25;
    if (t === "sphere" ? _ = 30 : t === "torus" ? _ = 35 : t === "cylinder" && (_ = 30), o >= 1)
      h.left = -_ * o, h.right = _ * o, h.top = _, h.bottom = -_;
    else {
      h.left = -_, h.right = _, h.top = _ / o, h.bottom = -_ / o;
      const m = 1.05;
      h.left *= m, h.right *= m, h.top *= m, h.bottom *= m;
    }
  }
  h.left /= a, h.right /= a, h.top /= a, h.bottom /= a, h.near = -100, h.far = 1e3, h.updateProjectionMatrix();
}
function ne(h, e, i, s) {
  const r = h / 2, t = e / 2, a = Math.floor(i), o = Math.floor(s), _ = a + 1, m = o + 1, w = h / a, v = e / o, u = [], x = [], l = [], f = [];
  for (let b = 0; b < m; b++) {
    const A = b * v - t;
    for (let y = 0; y < _; y++) {
      const c = y * w - r;
      x.push(c, -A, 0), l.push(0, 0, 1), f.push(y / a), f.push(1 - b / o);
    }
  }
  for (let b = 0; b < o; b++)
    for (let A = 0; A < a; A++) {
      const y = A + _ * b, c = A + _ * (b + 1), T = A + 1 + _ * (b + 1), F = A + 1 + _ * b;
      u.push(y, c, F), u.push(c, T, F);
    }
  const R = x.length / 3 > 65535, d = [];
  for (let b = 0; b < u.length; b += 3) {
    const A = u[b], y = u[b + 1], c = u[b + 2];
    d.push(A, y, y, c, c, A);
  }
  return {
    position: new Float32Array(x),
    normal: new Float32Array(l),
    uv: new Float32Array(f),
    index: R ? new Uint32Array(u) : new Uint16Array(u),
    wireframeIndex: R ? new Uint32Array(d) : new Uint16Array(d)
  };
}
function ae(h, e, i) {
  const s = [], r = [], t = [], a = [], o = Math.floor(e), _ = Math.floor(i);
  for (let v = 0; v <= _; v++) {
    const u = v / _, x = u * Math.PI;
    for (let l = 0; l <= o; l++) {
      const f = l / o, R = f * Math.PI * 2, d = -h * Math.sin(x) * Math.cos(R), b = h * Math.cos(x), A = h * Math.sin(x) * Math.sin(R);
      s.push(d, b, A);
      const y = Math.sqrt(d * d + b * b + A * A);
      r.push(d / y, b / y, A / y), t.push(f, 1 - u);
    }
  }
  for (let v = 0; v < _; v++)
    for (let u = 0; u < o; u++) {
      const x = u + (o + 1) * v, l = u + (o + 1) * (v + 1), f = u + 1 + (o + 1) * (v + 1), R = u + 1 + (o + 1) * v;
      a.push(x, l, R), a.push(l, f, R);
    }
  const m = s.length / 3 > 65535, w = [];
  for (let v = 0; v < a.length; v += 3) {
    const u = a[v], x = a[v + 1], l = a[v + 2];
    w.push(u, x, x, l, l, u);
  }
  return {
    position: new Float32Array(s),
    normal: new Float32Array(r),
    uv: new Float32Array(t),
    index: m ? new Uint32Array(a) : new Uint16Array(a),
    wireframeIndex: m ? new Uint32Array(w) : new Uint16Array(w)
  };
}
function le(h, e, i, s) {
  const r = [], t = [], a = [], o = [], _ = Math.floor(i), m = Math.floor(s);
  for (let u = 0; u <= _; u++) {
    const x = u / _ * Math.PI * 2;
    for (let l = 0; l <= m; l++) {
      const f = l / m * Math.PI * 2, R = (h + e * Math.cos(x)) * Math.cos(f), d = (h + e * Math.cos(x)) * Math.sin(f), b = e * Math.sin(x);
      r.push(R, d, b);
      const A = h * Math.cos(f), y = h * Math.sin(f), c = R - A, T = d - y, F = b, g = Math.sqrt(c * c + T * T + F * F);
      t.push(c / g, T / g, F / g), a.push(l / m, u / _);
    }
  }
  for (let u = 1; u <= _; u++)
    for (let x = 1; x <= m; x++) {
      const l = (m + 1) * u + x - 1, f = (m + 1) * (u - 1) + x - 1, R = (m + 1) * (u - 1) + x, d = (m + 1) * u + x;
      o.push(l, f, d), o.push(f, R, d);
    }
  const w = r.length / 3 > 65535, v = [];
  for (let u = 0; u < o.length; u += 3) {
    const x = o[u], l = o[u + 1], f = o[u + 2];
    v.push(x, l, l, f, f, x);
  }
  return {
    position: new Float32Array(r),
    normal: new Float32Array(t),
    uv: new Float32Array(a),
    index: w ? new Uint32Array(o) : new Uint16Array(o),
    wireframeIndex: w ? new Uint32Array(v) : new Uint16Array(v)
  };
}
function _e(h, e, i, s, r) {
  const t = [], a = [], o = [], _ = [], m = Math.floor(s), w = Math.floor(r), v = i / 2;
  for (let l = 0; l <= w; l++) {
    const f = l / w, R = f * i - v, d = f * (e - h) + h;
    for (let b = 0; b <= m; b++) {
      const A = b / m, y = A * Math.PI * 2, c = Math.sin(y), T = Math.cos(y);
      t.push(d * c, -R, d * T), a.push(c, 0, T), o.push(A, 1 - f);
    }
  }
  for (let l = 0; l < w; l++)
    for (let f = 0; f < m; f++) {
      const R = f + (m + 1) * l, d = f + (m + 1) * (l + 1), b = f + 1 + (m + 1) * (l + 1), A = f + 1 + (m + 1) * l;
      _.push(R, d, A), _.push(d, b, A);
    }
  const u = t.length / 3 > 65535, x = [];
  for (let l = 0; l < _.length; l += 3) {
    const f = _[l], R = _[l + 1], d = _[l + 2];
    x.push(f, R, R, d, d, f);
  }
  return {
    position: new Float32Array(t),
    normal: new Float32Array(a),
    uv: new Float32Array(o),
    index: u ? new Uint32Array(_) : new Uint16Array(_),
    wireframeIndex: u ? new Uint32Array(x) : new Uint16Array(x)
  };
}
function ce(h, e, i, s, r, t) {
  const a = h / 2, o = e / 2, _ = Math.floor(i), m = Math.floor(s), w = _ + 1, v = m + 1, u = h / _, x = e / m, l = [], f = [], R = [], d = [];
  for (let y = 0; y < v; y++) {
    const c = y * x - o;
    for (let T = 0; T < w; T++) {
      const F = T * u - a;
      let g = F, S = c, B = 0, E = 0, C = 0, U = 1;
      if (Math.abs(r) > 1e-3) {
        const z = h / r, M = F / z;
        g = z * Math.sin(M), B = z * (1 - Math.cos(M)), E = Math.sin(M), U = Math.cos(M);
      }
      if (Math.abs(t) > 1e-3) {
        const z = c / e * t, M = Math.cos(z), P = Math.sin(z), I = g * M - B * P, k = g * P + B * M;
        g = I, B = k;
        const D = E * M - U * P, W = E * P + U * M;
        E = D, U = W;
      }
      l.push(g, -S, B), f.push(E, C, U), R.push(T / _), R.push(1 - y / m);
    }
  }
  for (let y = 0; y < m; y++)
    for (let c = 0; c < _; c++) {
      const T = c + w * y, F = c + w * (y + 1), g = c + 1 + w * (y + 1), S = c + 1 + w * y;
      d.push(T, F, S), d.push(F, g, S);
    }
  const b = l.length / 3 > 65535, A = [];
  for (let y = 0; y < d.length; y += 3) {
    const c = d[y], T = d[y + 1], F = d[y + 2];
    A.push(c, T, T, F, F, c);
  }
  return {
    position: new Float32Array(l),
    normal: new Float32Array(f),
    uv: new Float32Array(R),
    index: b ? new Uint32Array(d) : new Uint16Array(d),
    wireframeIndex: b ? new Uint32Array(A) : new Uint16Array(A)
  };
}
const gt = {
  kty: "EC",
  crv: "P-256",
  x: "n9A9jNvLNR6QJaPP4ZdpbXtPFz3ASUfeeQm11Jd53Rg",
  y: "EoG5ezJ3hr4c62JjpsyabotdFeU-A1LyH-qHyabnKc0",
  key_ops: ["verify"],
  ext: !0
};
function ue(h) {
  let e = h.replace(/-/g, "+").replace(/_/g, "/");
  for (; e.length % 4 !== 0; )
    e += "=";
  const i = atob(e), s = new Uint8Array(i.length);
  for (let r = 0; r < i.length; r++)
    s[r] = i.charCodeAt(r);
  return s;
}
function yt(h) {
  if (typeof window > "u" || !window.location)
    return !0;
  const e = window.location.hostname.toLowerCase(), i = h.toLowerCase();
  return !!(e === "localhost" || e === "127.0.0.1" || e === "0.0.0.0" || e === "[::1]" || e.endsWith(".localhost") || e === i || e.endsWith("." + i));
}
async function xt(h) {
  try {
    if (typeof crypto > "u" || !crypto.subtle || typeof crypto.subtle.verify != "function")
      return { valid: !1 };
    if (!h.startsWith("NEAT-"))
      return { valid: !1 };
    const e = h.slice(5), i = e.indexOf(".");
    if (i === -1)
      return { valid: !1 };
    const s = e.slice(0, i), r = e.slice(i + 1);
    if (!s || !r)
      return { valid: !1 };
    const t = ue(s), a = new TextDecoder().decode(t), o = JSON.parse(a);
    if (!o.domain || typeof o.domain != "string")
      return { valid: !1 };
    if (!yt(o.domain))
      return { valid: !1 };
    const _ = ue(r), m = await crypto.subtle.importKey(
      "jwk",
      gt,
      { name: "ECDSA", namedCurve: "P-256" },
      !1,
      ["verify"]
    );
    return await crypto.subtle.verify(
      { name: "ECDSA", hash: "SHA-256" },
      m,
      _,
      t
    ) ? { valid: !0, payload: o } : { valid: !1 };
  } catch {
    return { valid: !1 };
  }
}
function bt() {
  console.info(
    `%c\u{1F308} Neat Gradients%c

Licensed under MIT + The Commons Clause.
Free for personal and commercial use.
Selling this software or its derivatives is strictly prohibited.
Get a license key to remove the watermark and this message: https://neat.firecms.co`,
    "font-weight: bold; font-size: 14px; color: #FF5772;",
    "color: inherit;"
  );
}
const Y = 50, V = 80, $ = 6, wt = [
  ["speed", "_speed", 20, 1 / 20, "u"],
  ["horizontalPressure", "_horizontalPressure", 4, 1 / 4, "u"],
  ["verticalPressure", "_verticalPressure", 4, 1 / 4, "u"],
  ["waveFrequencyX", "_waveFrequencyX", 1 / 0.04, 0.04, "u"],
  ["waveFrequencyY", "_waveFrequencyY", 1 / 0.04, 0.04, "u"],
  ["waveAmplitude", "_waveAmplitude", 1 / 0.75, 0.75, "u"],
  ["highlights", "_highlights", 100, 1 / 100, "u"],
  ["shadows", "_shadows", 100, 1 / 100, "u"],
  ["colorSaturation", "_saturation", 10, 1 / 10, "u"],
  ["colorBlending", "_colorBlending", 10, 1 / 10, "u"],
  ["yOffsetWaveMultiplier", "_yOffsetWaveMultiplier", 1e3, 1 / 1e3, "u"],
  ["yOffsetColorMultiplier", "_yOffsetColorMultiplier", 1e3, 1 / 1e3, "u"],
  ["yOffsetFlowMultiplier", "_yOffsetFlowMultiplier", 1e3, 1 / 1e3, "u"],
  ["colorBrightness", "_brightness", 1, 1, "u"],
  ["grainIntensity", "_grainIntensity", 1, 1, "u"],
  ["grainSparsity", "_grainSparsity", 1, 1, "u"],
  ["grainSpeed", "_grainSpeed", 1, 1, "u"],
  ["wireframe", "_wireframe", 1, 1, "u"],
  ["backgroundAlpha", "_backgroundAlpha", 1, 1, "u"],
  ["flowDistortionA", "_flowDistortionA", 1, 1, "u"],
  ["flowDistortionB", "_flowDistortionB", 1, 1, "u"],
  ["flowScale", "_flowScale", 1, 1, "u"],
  ["flowEase", "_flowEase", 1, 1, "u"],
  ["flowEnabled", "_flowEnabled", 1, 1, "u"],
  ["textureEase", "_textureEase", 1, 1, "u"],
  ["silhouetteFade", "_silhouetteFade", 1, 1, "u"],
  ["cylinderFade", "_cylinderFade", 1, 1, "u"],
  ["ribbonFade", "_ribbonFade", 1, 1, "u"],
  ["flatShading", "_flatShading", 1, 1, "u"],
  ["domainWarpEnabled", "_domainWarpEnabled", 1, 1, "u"],
  ["domainWarpIntensity", "_domainWarpIntensity", 1, 1, "u"],
  ["domainWarpScale", "_domainWarpScale", 1, 1, "u"],
  ["vignetteIntensity", "_vignetteIntensity", 1, 1, "u"],
  ["vignetteRadius", "_vignetteRadius", 1, 1, "u"],
  ["fresnelEnabled", "_fresnelEnabled", 1, 1, "u"],
  ["fresnelPower", "_fresnelPower", 1, 1, "u"],
  ["fresnelIntensity", "_fresnelIntensity", 1, 1, "u"],
  ["iridescenceEnabled", "_iridescenceEnabled", 1, 1, "u"],
  ["iridescenceIntensity", "_iridescenceIntensity", 1, 1, "u"],
  ["iridescenceSpeed", "_iridescenceSpeed", 1, 1, "u"],
  ["bloomIntensity", "_bloomIntensity", 1, 1, "u"],
  ["bloomThreshold", "_bloomThreshold", 1, 1, "u"],
  ["chromaticAberration", "_chromaticAberration", 1, 1, "u"],
  ["shapeRotationX", "_shapeRotationX", 1, 1, "u"],
  ["shapeRotationY", "_shapeRotationY", 1, 1, "u"],
  ["shapeRotationZ", "_shapeRotationZ", 1, 1, "u"],
  ["shapeAutoRotateSpeedX", "_shapeAutoRotateSpeedX", 1, 1, "u"],
  ["shapeAutoRotateSpeedY", "_shapeAutoRotateSpeedY", 1, 1, "u"],
  ["cameraX", "_cameraX", 1, 1, "u"],
  ["cameraY", "_cameraY", 1, 1, "u"],
  ["cameraZ", "_cameraZ", 1, 1, "u"],
  ["cameraRotationX", "_cameraRotationX", 1, 1, "u"],
  ["cameraRotationY", "_cameraRotationY", 1, 1, "u"],
  ["cameraRotationZ", "_cameraRotationZ", 1, 1, "u"],
  ["textureVoidLikelihood", "_textureVoidLikelihood", 1, 1, "t"],
  ["textureVoidWidthMin", "_textureVoidWidthMin", 1, 1, "t"],
  ["textureVoidWidthMax", "_textureVoidWidthMax", 1, 1, "t"],
  ["textureBandDensity", "_textureBandDensity", 1, 1, "t"],
  ["textureColorBlending", "_textureColorBlending", 1, 1, "t"],
  ["textureSeed", "_textureSeed", 1, 1, "t"],
  ["transparentTextureVoid", "_transparentTextureVoid", 1, 1, "t"],
  ["proceduralBackgroundColor", "_proceduralBackgroundColor", 1, 1, "t"],
  ["textureShapeTriangles", "_textureShapeTriangles", 1, 1, "t"],
  ["textureShapeCircles", "_textureShapeCircles", 1, 1, "t"],
  ["textureShapeBars", "_textureShapeBars", 1, 1, "t"],
  ["textureShapeSquiggles", "_textureShapeSquiggles", 1, 1, "t"],
  ["sphereRadius", "_sphereRadius", 1, 1, "g"],
  ["torusRadius", "_torusRadius", 1, 1, "g"],
  ["torusTube", "_torusTube", 1, 1, "g"],
  ["cylinderRadius", "_cylinderRadius", 1, 1, "g"],
  ["cylinderHeight", "_cylinderHeight", 1, 1, "g"],
  ["planeBend", "_planeBend", 1, 1, "g"],
  ["planeTwist", "_planeTwist", 1, 1, "g"]
];
class Rt {
  _ref;
  _licensed = !1;
  _speed = -1;
  _horizontalPressure = -1;
  _verticalPressure = -1;
  _waveFrequencyX = -1;
  _waveFrequencyY = -1;
  _waveAmplitude = -1;
  _shadows = -1;
  _highlights = -1;
  _saturation = -1;
  _brightness = -1;
  _grainScale = -1;
  _grainIntensity = -1;
  _grainSparsity = -1;
  _grainSpeed = -1;
  _colorBlending = -1;
  _resolution = 1;
  _colors = [];
  _wireframe = !1;
  _backgroundColor = "#FFFFFF";
  _backgroundColorRgb = [1, 1, 1];
  _backgroundAlpha = 1;
  _flowDistortionA = 0;
  _flowDistortionB = 0;
  _flowScale = 1;
  _flowEase = 0;
  _flowEnabled = !0;
  glState;
  _enableProceduralTexture = !1;
  _textureVoidLikelihood = 0.45;
  _textureVoidWidthMin = 200;
  _textureVoidWidthMax = 486;
  _textureBandDensity = 2.15;
  _textureColorBlending = 0.01;
  _textureSeed = 333;
  _textureEase = 0.5;
  _transparentTextureVoid = !1;
  _domainWarpEnabled = !1;
  _domainWarpIntensity = 0.5;
  _domainWarpScale = 1;
  _vignetteIntensity = 0.5;
  _vignetteRadius = 0.8;
  _fresnelEnabled = !1;
  _fresnelPower = 2;
  _fresnelIntensity = 0.5;
  _fresnelColor = "#FFFFFF";
  _fresnelColorRgb = [1, 1, 1];
  _iridescenceEnabled = !1;
  _iridescenceIntensity = 0.5;
  _iridescenceSpeed = 1;
  _bloomIntensity = 0;
  _bloomThreshold = 0.7;
  _chromaticAberration = 0;
  _silhouetteFade = 0.25;
  _cylinderFade = 0.08;
  _ribbonFade = 0.05;
  _flatShading = !0;
  _shapeType = "plane";
  _shapeRotationX = 0;
  _shapeRotationY = 0;
  _shapeRotationZ = 0;
  _shapeAutoRotateSpeedX = 0;
  _shapeAutoRotateSpeedY = 0;
  _sphereRadius = 15;
  _torusRadius = 15;
  _torusTube = 5;
  _cylinderRadius = 10;
  _cylinderHeight = 40;
  _planeBend = 0;
  _planeTwist = 0;
  _cameraLock = !1;
  _cameraX = 0;
  _cameraY = 0;
  _cameraZ = 0;
  _cameraRotationX = 0;
  _cameraRotationY = 0;
  _cameraRotationZ = 0;
  _cameraZoom = 1;
  _proceduralTexture = null;
  _proceduralBackgroundColor = "#000000";
  _textureShapeTriangles = 20;
  _textureShapeCircles = 15;
  _textureShapeBars = 15;
  _textureShapeSquiggles = 10;
  requestRef = -1;
  sizeObserver;
  _initialized = !1;
  _cachedColorRgb = [];
  _yOffset = 0;
  _yOffsetWaveMultiplier = 4e-3;
  _yOffsetColorMultiplier = 4e-3;
  _yOffsetFlowMultiplier = 4e-3;
  _sourceCanvas = null;
  _sourceCtx = null;
  _maskedCanvas = null;
  _maskedCtx = null;
  _resizeTimeoutId = null;
  _textureNeedsUpdate = !1;
  _colorsChanged = !0;
  _uniformsDirty = !0;
  _textureDirty = !0;
  _yOffsetDirty = !1;
  _modelViewMatrix = new fe();
  _isVisible = !0;
  _visibilityObserver = null;
  _visibilityHandler = null;
  _watermarkProgram = null;
  _watermarkTexture = null;
  _watermarkBuffer = null;
  _watermarkTexCoordBuffer = null;
  _watermarkWidth = 0;
  _watermarkHeight = 0;
  _watermarkMargin = 4;
  _wmLocPos = -1;
  _wmLocTc = -1;
  _wmLocTex = null;
  _wmPosData = new Float32Array(8);
  _wmClickHandler = null;
  _wmMoveHandler = null;
  _wmMoveRafPending = !1;
  _wmCachedRect = null;
  _wmRectCacheTime = 0;
  _gradientVAO = null;
  _watermarkVAO = null;
  constructor(e) {
    const {
      ref: i,
      speed: s = 4,
      horizontalPressure: r = 3,
      verticalPressure: t = 3,
      waveFrequencyX: a = 5,
      waveFrequencyY: o = 5,
      waveAmplitude: _ = 3,
      colors: m,
      highlights: w = 4,
      shadows: v = 4,
      colorSaturation: u = 0,
      colorBrightness: x = 1,
      colorBlending: l = 5,
      grainScale: f = 2,
      grainIntensity: R = 0.55,
      grainSparsity: d = 0,
      grainSpeed: b = 0.1,
      wireframe: A = !1,
      backgroundColor: y = "#FFFFFF",
      backgroundAlpha: c = 1,
      resolution: T = 1,
      seed: F,
      yOffset: g = 0,
      yOffsetWaveMultiplier: S = 4,
      yOffsetColorMultiplier: B = 4,
      yOffsetFlowMultiplier: E = 4,
      flowDistortionA: C = 0,
      flowDistortionB: U = 0,
      flowScale: z = 1,
      flowEase: M = 0,
      flowEnabled: P = !0,
      enableProceduralTexture: I = !1,
      textureVoidLikelihood: k = 0.45,
      textureVoidWidthMin: D = 200,
      textureVoidWidthMax: W = 486,
      textureBandDensity: he = 2.15,
      textureColorBlending: de = 0.01,
      textureSeed: me = 333,
      textureEase: pe = 0.5,
      proceduralBackgroundColor: ve = "#000000",
      transparentTextureVoid: ge = !1,
      textureShapeTriangles: ye = 20,
      textureShapeCircles: xe = 15,
      textureShapeBars: be = 15,
      textureShapeSquiggles: we = 10,
      domainWarpEnabled: Re = !1,
      domainWarpIntensity: Ae = 0.5,
      domainWarpScale: Se = 1,
      vignetteIntensity: Te = 0,
      vignetteRadius: Ee = 0.8,
      fresnelEnabled: Fe = !1,
      fresnelPower: Pe = 2,
      fresnelIntensity: Ce = 0.5,
      fresnelColor: Me = "#FFFFFF",
      iridescenceEnabled: Be = !1,
      iridescenceIntensity: Ue = 0.5,
      iridescenceSpeed: ze = 1,
      bloomIntensity: Ie = 0,
      bloomThreshold: Le = 0.7,
      chromaticAberration: De = 0,
      silhouetteFade: Oe = 0.25,
      cylinderFade: ke = 0.08,
      ribbonFade: Ye = 0.05,
      flatShading: Ve = !0,
      cameraLock: We = !1,
      cameraX: Ne = 0,
      cameraY: Xe = 0,
      cameraZ: He = 0,
      cameraRotationX: qe = 0,
      cameraRotationY: Ge = 0,
      cameraRotationZ: Ze = 0,
      cameraZoom: je = 1,
      shapeType: $e = "plane",
      shapeRotationX: Ke = 0,
      shapeRotationY: Je = 0,
      shapeRotationZ: Qe = 0,
      shapeAutoRotateSpeedX: et = 0,
      shapeAutoRotateSpeedY: tt = 0,
      sphereRadius: it = 15,
      torusRadius: ot = 15,
      torusTube: rt = 5,
      cylinderRadius: st = 10,
      cylinderHeight: nt = 40,
      planeBend: at = 0,
      planeTwist: lt = 0,
      licenseKey: Q,
      preserveDrawingBuffer: _t = !1
    } = e;
    this._ref = i, this.destroy = this.destroy.bind(this), this._initScene = this._initScene.bind(this), this.speed = s, this.horizontalPressure = r, this.verticalPressure = t, this.waveFrequencyX = a, this.waveFrequencyY = o, this.waveAmplitude = _, this.colorBlending = l, this._resolution = T, this.grainScale = f, this.grainIntensity = R, this.grainSparsity = d, this.grainSpeed = b, this.colors = m, this.shadows = v, this.highlights = w, this.colorSaturation = u, this.colorBrightness = x, this.wireframe = A, this.backgroundColor = y, this.backgroundAlpha = c, this.yOffset = g, this.yOffsetWaveMultiplier = S, this.yOffsetColorMultiplier = B, this.yOffsetFlowMultiplier = E, this.flowDistortionA = C, this.flowDistortionB = U, this.flowScale = z, this.flowEase = M, this.flowEnabled = P, this.enableProceduralTexture = I, this.textureVoidLikelihood = k, this.textureVoidWidthMin = D, this.textureVoidWidthMax = W, this.textureBandDensity = he, this.textureColorBlending = de, this.textureSeed = me, this.textureEase = pe, this._proceduralBackgroundColor = ve, this.transparentTextureVoid = ge, this._textureShapeTriangles = ye, this._textureShapeCircles = xe, this._textureShapeBars = be, this._textureShapeSquiggles = we, this.domainWarpEnabled = Re, this.domainWarpIntensity = Ae, this.domainWarpScale = Se, this.vignetteIntensity = Te, this.vignetteRadius = Ee, this.fresnelEnabled = Fe, this.fresnelPower = Pe, this.fresnelIntensity = Ce, this.fresnelColor = Me, this.iridescenceEnabled = Be, this.iridescenceIntensity = Ue, this.iridescenceSpeed = ze, this.bloomIntensity = Ie, this.bloomThreshold = Le, this.chromaticAberration = De, this.silhouetteFade = Oe, this.cylinderFade = ke, this.ribbonFade = Ye, this._flatShading = Ve, this._cameraLock = We, this._cameraX = Ne, this._cameraY = Xe, this._cameraZ = He, this._cameraRotationX = qe, this._cameraRotationY = Ge, this._cameraRotationZ = Ze, this._cameraZoom = je, this._shapeType = $e, this._shapeRotationX = Ke, this._shapeRotationY = Je, this._shapeRotationZ = Qe, this._shapeAutoRotateSpeedX = et, this._shapeAutoRotateSpeedY = tt, this._sphereRadius = it, this._torusRadius = ot, this._torusTube = rt, this._cylinderRadius = st, this._cylinderHeight = nt, this._planeBend = at, this._planeTwist = lt, this.glState = this._initScene(T, _t), this._initWatermark(), St(), Q ? xt(Q).then((n) => {
      this._licensed = n.valid;
    }) : bt();
    let q = F !== void 0 ? F : At(), G = performance.now();
    const Z = () => {
      const { gl: n, program: N, locations: p, indexCount: X, indexType: H } = this.glState;
      if (this._initialized) {
        const ee = performance.now();
        q += (ee - G) / 1e3 * this._speed, G = ee, n.useProgram(N), n.uniform1f(p.uniforms.u_time, q);
        const K = this.glState.camera, O = this._modelViewMatrix;
        O.identity(), O.translate(
          -K.position[0] - this._cameraX,
          -K.position[1] - this._cameraY,
          -K.position[2] - this._cameraZ
        ), O.translate(0, 0, -1), O.rotateX(-this._cameraRotationX), O.rotateY(-this._cameraRotationY), O.rotateZ(-this._cameraRotationZ);
        let J = this._shapeRotationX, te = this._shapeRotationY, ut = this._shapeRotationZ;
        this._shapeAutoRotateSpeedX !== 0 && (J += q * this._shapeAutoRotateSpeedX * 0.1), this._shapeAutoRotateSpeedY !== 0 && (te += q * this._shapeAutoRotateSpeedY * 0.1), this._shapeType === "plane" || this._shapeType === "ribbon" ? O.rotateX(J - Math.PI / 3.5) : O.rotateX(J), O.rotateY(te), O.rotateZ(ut);
        const ie = p.uniforms.modelViewMatrix;
        if (ie && n.uniformMatrix4fv(ie, !1, O.elements), this._yOffsetDirty && !this._uniformsDirty && (n.uniform1f(p.uniforms.u_y_offset, this._yOffset), this._yOffsetDirty = !1), this._uniformsDirty) {
          n.uniform2f(p.uniforms.u_resolution, this._ref.clientWidth, this._ref.clientHeight), n.uniform2f(p.uniforms.u_color_pressure, this._horizontalPressure, this._verticalPressure), n.uniform1f(p.uniforms.u_wave_frequency_x, this._waveFrequencyX), n.uniform1f(p.uniforms.u_wave_frequency_y, this._waveFrequencyY), n.uniform1f(p.uniforms.u_wave_amplitude, this._waveAmplitude), n.uniform1f(p.uniforms.u_color_blending, this._colorBlending), n.uniform1f(p.uniforms.u_shadows, this._shadows), n.uniform1f(p.uniforms.u_highlights, this._highlights), n.uniform1f(p.uniforms.u_saturation, this._saturation), n.uniform1f(p.uniforms.u_brightness, this._brightness), n.uniform1f(p.uniforms.u_grain_intensity, this._grainIntensity), n.uniform1f(p.uniforms.u_grain_sparsity, this._grainSparsity), n.uniform1f(p.uniforms.u_grain_speed, this._grainSpeed), n.uniform1f(p.uniforms.u_grain_scale, this._grainScale), n.uniform1f(p.uniforms.u_y_offset, this._yOffset), n.uniform1f(p.uniforms.u_y_offset_wave_multiplier, this._yOffsetWaveMultiplier), n.uniform1f(p.uniforms.u_y_offset_color_multiplier, this._yOffsetColorMultiplier), n.uniform1f(p.uniforms.u_y_offset_flow_multiplier, this._yOffsetFlowMultiplier), n.uniform1f(p.uniforms.u_flow_distortion_a, this._flowDistortionA), n.uniform1f(p.uniforms.u_flow_distortion_b, this._flowDistortionB), n.uniform1f(p.uniforms.u_flow_scale, this._flowScale), n.uniform1f(p.uniforms.u_flow_ease, this._flowEase), n.uniform1f(p.uniforms.u_flow_enabled, this._flowEnabled ? 1 : 0);
          let L = 0;
          this._shapeType === "sphere" ? L = 1 : this._shapeType === "torus" ? L = 2 : this._shapeType === "cylinder" ? L = 3 : this._shapeType === "ribbon" && (L = 4), n.uniform1f(p.uniforms.u_shape_type, L), n.uniform1f(p.uniforms.u_enable_procedural_texture, this._enableProceduralTexture ? 1 : 0), n.uniform1f(p.uniforms.u_texture_ease, this._textureEase), n.uniform1f(p.uniforms.u_transparent_texture_void, this._transparentTextureVoid ? 1 : 0), n.uniform1f(p.uniforms.u_domain_warp_enabled, this._domainWarpEnabled ? 1 : 0), n.uniform1f(p.uniforms.u_domain_warp_intensity, this._domainWarpIntensity), n.uniform1f(p.uniforms.u_domain_warp_scale, this._domainWarpScale), n.uniform1f(p.uniforms.u_vignette_intensity, this._vignetteIntensity), n.uniform1f(p.uniforms.u_vignette_radius, this._vignetteRadius), n.uniform1f(p.uniforms.u_fresnel_enabled, this._fresnelEnabled ? 1 : 0), n.uniform1f(p.uniforms.u_fresnel_power, this._fresnelPower), n.uniform1f(p.uniforms.u_fresnel_intensity, this._fresnelIntensity), n.uniform3fv(p.uniforms.u_fresnel_color, this._fresnelColorRgb), n.uniform1f(p.uniforms.u_iridescence_enabled, this._iridescenceEnabled ? 1 : 0), n.uniform1f(p.uniforms.u_iridescence_intensity, this._iridescenceIntensity), n.uniform1f(p.uniforms.u_iridescence_speed, this._iridescenceSpeed), n.uniform1f(p.uniforms.u_bloom_intensity, this._bloomIntensity), n.uniform1f(p.uniforms.u_bloom_threshold, this._bloomThreshold), n.uniform1f(p.uniforms.u_chromatic_aberration, this._chromaticAberration), n.uniform1f(p.uniforms.u_silhouette_fade, this._silhouetteFade), n.uniform1f(p.uniforms.u_cylinder_fade, this._cylinderFade), n.uniform1f(p.uniforms.u_ribbon_fade, this._ribbonFade), n.uniform1f(p.uniforms.u_flat_shading, this._flatShading ? 1 : 0), this._uniformsDirty = !1, this._yOffsetDirty = !1;
        }
        if (this._textureNeedsUpdate && this._enableProceduralTexture && (this._proceduralTexture && n.deleteTexture(this._proceduralTexture), this._proceduralTexture = this._createProceduralTexture(n), this._textureNeedsUpdate = !1, this._textureDirty = !0), this._textureDirty && this._proceduralTexture && (n.activeTexture(n.TEXTURE1), n.bindTexture(n.TEXTURE_2D, this._proceduralTexture), n.uniform1i(p.uniforms.u_procedural_texture, 1), this._textureDirty = !1), this._colorsChanged) {
          this._colorsChanged = !1;
          for (let L = 0; L < $; L++)
            if (L < this._colors.length) {
              const oe = this._colors[L], ft = this._cachedColorRgb[L] || [0, 0, 0];
              n.uniform1f(p.uniforms[`u_colors[${L}].is_active`], oe.enabled ? 1 : 0), n.uniform3fv(p.uniforms[`u_colors[${L}].color`], ft), n.uniform1f(p.uniforms[`u_colors[${L}].influence`], oe.influence || 0);
            } else
              n.uniform1f(p.uniforms[`u_colors[${L}].is_active`], 0);
          n.uniform1i(p.uniforms.u_colors_count, $);
        }
      }
      n.clearColor(
        this._backgroundColorRgb[0],
        this._backgroundColorRgb[1],
        this._backgroundColorRgb[2],
        this._backgroundAlpha
      ), n.clear(n.COLOR_BUFFER_BIT | n.DEPTH_BUFFER_BIT), this._wireframe ? (n.bindBuffer(n.ELEMENT_ARRAY_BUFFER, this.glState.buffers.wireframeIndex), n.drawElements(n.LINES, this.glState.wireframeIndexCount, H, 0), n.bindBuffer(n.ELEMENT_ARRAY_BUFFER, this.glState.buffers.index)) : n.drawElements(n.TRIANGLES, X, H, 0), this._licensed || this._renderWatermark(n), this._isVisible && (this.requestRef = requestAnimationFrame(Z));
    };
    this._visibilityObserver = new IntersectionObserver((n) => {
      const N = this._isVisible;
      this._isVisible = n[0].isIntersecting && document.visibilityState !== "hidden", this._isVisible && !N && (G = performance.now(), this.requestRef = requestAnimationFrame(Z));
    }, { threshold: 0 }), this._visibilityObserver.observe(i), this._visibilityHandler = () => {
      const n = this._isVisible;
      document.visibilityState === "hidden" ? this._isVisible = !1 : (this._isVisible = !0, n || (G = performance.now(), this.requestRef = requestAnimationFrame(Z)));
    }, document.addEventListener("visibilitychange", this._visibilityHandler);
    const ct = () => {
      const { gl: n, camera: N } = this.glState, p = this._ref.clientWidth, X = this._ref.clientHeight;
      this._ref.width = p, this._ref.height = X, n.viewport(0, 0, p, X), j(N, p, X, Y, V, this._shapeType, this._cameraZoom);
      const H = this.glState.locations.uniforms.projectionMatrix;
      n.useProgram(this.glState.program), H && n.uniformMatrix4fv(H, !1, N.projectionMatrix.elements);
    };
    this.sizeObserver = new ResizeObserver(() => {
      this._resizeTimeoutId !== null && clearTimeout(this._resizeTimeoutId), this._resizeTimeoutId = window.setTimeout(() => {
        ct(), this._resizeTimeoutId = null;
      }, 100);
    }), this.sizeObserver.observe(i), Z();
  }
  destroy() {
    if (cancelAnimationFrame(this.requestRef), this.sizeObserver.disconnect(), this._visibilityObserver && (this._visibilityObserver.disconnect(), this._visibilityObserver = null), this._visibilityHandler && (document.removeEventListener("visibilitychange", this._visibilityHandler), this._visibilityHandler = null), this._resizeTimeoutId !== null && (clearTimeout(this._resizeTimeoutId), this._resizeTimeoutId = null), this._wmClickHandler && (document.removeEventListener("click", this._wmClickHandler, !0), this._wmClickHandler = null), this._wmMoveHandler && (document.removeEventListener("mousemove", this._wmMoveHandler), this._wmMoveHandler = null), this.glState) {
      const e = this.glState.gl;
      e.deleteProgram(this.glState.program), e.deleteBuffer(this.glState.buffers.position), e.deleteBuffer(this.glState.buffers.normal), e.deleteBuffer(this.glState.buffers.uv), e.deleteBuffer(this.glState.buffers.index), e.deleteBuffer(this.glState.buffers.wireframeIndex), this._watermarkProgram && e.deleteProgram(this._watermarkProgram), this._watermarkTexture && e.deleteTexture(this._watermarkTexture), this._watermarkBuffer && e.deleteBuffer(this._watermarkBuffer), this._watermarkTexCoordBuffer && e.deleteBuffer(this._watermarkTexCoordBuffer);
      const i = e;
      i.deleteVertexArray && (this._gradientVAO && i.deleteVertexArray(this._gradientVAO), this._watermarkVAO && i.deleteVertexArray(this._watermarkVAO));
    }
    this._proceduralTexture && this.glState && this.glState.gl.deleteTexture(this._proceduralTexture);
  }
  get colors() {
    return this._colors;
  }
  set colors(e) {
    this._uniformsDirty = !0, this._colors = e, this._cachedColorRgb = e.map((i) => this._hexToRgb(i.color)), this._colorsChanged = !0;
  }
  get grainScale() {
    return this._grainScale;
  }
  set grainScale(e) {
    this._uniformsDirty = !0, this._grainScale = e == 0 ? 1 : e;
  }
  get resolution() {
    return this._resolution;
  }
  set resolution(e) {
    this._resolution !== e && (this._resolution = e, this._updateGeometry());
  }
  get backgroundColor() {
    return this._backgroundColor;
  }
  set backgroundColor(e) {
    this._uniformsDirty = !0, this._backgroundColor = e, this._backgroundColorRgb = this._hexToRgb(e);
  }
  get yOffset() {
    return this._yOffset;
  }
  set yOffset(e) {
    this._yOffset !== e && (this._yOffsetDirty = !0, this._yOffset = e);
  }
  get enableProceduralTexture() {
    return this._enableProceduralTexture;
  }
  set enableProceduralTexture(e) {
    this._uniformsDirty = !0, this._enableProceduralTexture = e, e && !this._proceduralTexture && (this._textureNeedsUpdate = !0);
  }
  _updateGeometry() {
    if (!this.glState)
      return;
    const e = this.glState.gl, i = this._resolution || 1;
    let s;
    this._shapeType === "sphere" ? s = ae(this._sphereRadius, 120 * i, 120 * i) : this._shapeType === "torus" ? s = le(this._torusRadius, this._torusTube, 120 * i, 120 * i) : this._shapeType === "cylinder" ? s = _e(this._cylinderRadius, this._cylinderRadius, this._cylinderHeight, 120 * i, 120 * i) : this._shapeType === "ribbon" ? s = ce(Y, V, 240 * i, 240 * i, this._planeBend, this._planeTwist) : s = ne(Y, V, 240 * i, 240 * i);
    const { position: r, normal: t, uv: a, index: o, wireframeIndex: _ } = s;
    e.bindBuffer(e.ARRAY_BUFFER, this.glState.buffers.position), e.bufferData(e.ARRAY_BUFFER, r, e.STATIC_DRAW), e.bindBuffer(e.ARRAY_BUFFER, this.glState.buffers.normal), e.bufferData(e.ARRAY_BUFFER, t, e.STATIC_DRAW), e.bindBuffer(e.ARRAY_BUFFER, this.glState.buffers.uv), e.bufferData(e.ARRAY_BUFFER, a, e.STATIC_DRAW), e.bindBuffer(e.ELEMENT_ARRAY_BUFFER, this.glState.buffers.index), e.bufferData(e.ELEMENT_ARRAY_BUFFER, o, e.STATIC_DRAW), e.bindBuffer(e.ELEMENT_ARRAY_BUFFER, this.glState.buffers.wireframeIndex), e.bufferData(e.ELEMENT_ARRAY_BUFFER, _, e.STATIC_DRAW), e.bindBuffer(e.ELEMENT_ARRAY_BUFFER, this.glState.buffers.index), this.glState.indexCount = o.length, this.glState.wireframeIndexCount = _.length, this.glState.indexType = o instanceof Uint32Array ? e.UNSIGNED_INT : e.UNSIGNED_SHORT;
    const m = this._ref.clientWidth, w = this._ref.clientHeight;
    j(this.glState.camera, m, w, Y, V, this._shapeType, this._cameraZoom);
    const v = this.glState.locations.uniforms.projectionMatrix;
    e.useProgram(this.glState.program), v && e.uniformMatrix4fv(v, !1, this.glState.camera.projectionMatrix.elements), this._uniformsDirty = !0;
  }
  _hexToRgb(e) {
    const i = parseInt(e.replace("#", ""), 16);
    return [
      (i >> 16 & 255) / 255,
      (i >> 8 & 255) / 255,
      (i & 255) / 255
    ];
  }
  _initScene(e, i = !1) {
    const s = this._ref.clientWidth, r = this._ref.clientHeight, t = this._ref.getContext("webgl2", { alpha: !0, preserveDrawingBuffer: i, antialias: !0 }) || this._ref.getContext("webgl", { alpha: !0, preserveDrawingBuffer: i, antialias: !0 });
    if (!t)
      throw new Error("WebGL not supported");
    t.getExtension("OES_standard_derivatives"), t.getExtension("OES_element_index_uint"), t.viewport(0, 0, s, r);
    let a;
    this._shapeType === "sphere" ? a = ae(this._sphereRadius, 120 * e, 120 * e) : this._shapeType === "torus" ? a = le(this._torusRadius, this._torusTube, 120 * e, 120 * e) : this._shapeType === "cylinder" ? a = _e(this._cylinderRadius, this._cylinderRadius, this._cylinderHeight, 120 * e, 120 * e) : this._shapeType === "ribbon" ? a = ce(Y, V, 240 * e, 240 * e, this._planeBend, this._planeTwist) : a = ne(Y, V, 240 * e, 240 * e);
    const { position: o, normal: _, uv: m, index: w, wireframeIndex: v } = a, u = t.createBuffer();
    t.bindBuffer(t.ARRAY_BUFFER, u), t.bufferData(t.ARRAY_BUFFER, o, t.STATIC_DRAW);
    const x = t.createBuffer();
    t.bindBuffer(t.ARRAY_BUFFER, x), t.bufferData(t.ARRAY_BUFFER, _, t.STATIC_DRAW);
    const l = t.createBuffer();
    t.bindBuffer(t.ARRAY_BUFFER, l), t.bufferData(t.ARRAY_BUFFER, m, t.STATIC_DRAW);
    const f = t.createBuffer();
    t.bindBuffer(t.ELEMENT_ARRAY_BUFFER, f), t.bufferData(t.ELEMENT_ARRAY_BUFFER, w, t.STATIC_DRAW);
    const R = t.createBuffer();
    t.bindBuffer(t.ELEMENT_ARRAY_BUFFER, R), t.bufferData(t.ELEMENT_ARRAY_BUFFER, v, t.STATIC_DRAW), t.bindBuffer(t.ELEMENT_ARRAY_BUFFER, f);
    const d = mt() + `
` + re() + `
` + se() + `
` + ht, b = t.createShader(t.VERTEX_SHADER);
    t.shaderSource(b, d), t.compileShader(b), t.getShaderParameter(b, t.COMPILE_STATUS) || (console.log("VERTEX_SHADER_ERROR_START"), console.log("Vertex shader error: ", t.getShaderInfoLog(b)), console.log("GL Error Code:", t.getError()), console.log("Vertex Shader Source Dump:"), console.log(d.split(`
`).map((P, I) => `${I + 1}: ${P}`).join(`
`)), console.log("VERTEX_SHADER_ERROR_END"));
    const A = pt() + `
` + se() + `
` + re() + `
` + dt, y = t.createShader(t.FRAGMENT_SHADER);
    t.shaderSource(y, A), t.compileShader(y), t.getShaderParameter(y, t.COMPILE_STATUS) || (console.log("FRAGMENT_SHADER_ERROR_START"), console.log("Fragment shader error: ", t.getShaderInfoLog(y)), console.log("GL Error Code:", t.getError()), console.log("Fragment Shader Source Dump:"), console.log(A.split(`
`).map((P, I) => `${I + 1}: ${P}`).join(`
`)), console.log("FRAGMENT_SHADER_ERROR_END"));
    const c = t.createProgram();
    t.attachShader(c, b), t.attachShader(c, y), t.linkProgram(c), t.getProgramParameter(c, t.LINK_STATUS) || (console.log("PROGRAM_LINK_ERROR_START"), console.log("Program linking error: ", t.getProgramInfoLog(c)), console.log("GL Error Code:", t.getError()), console.log("PROGRAM_LINK_ERROR_END")), t.useProgram(c);
    const T = new vt(0, 0, 0, 0, 0, 1e3);
    T.position = [0, 0, 5], j(T, s, r, Y, V, this._shapeType, this._cameraZoom);
    const F = t.getAttribLocation(c, "position"), g = t.getAttribLocation(c, "normal"), S = t.getAttribLocation(c, "uv");
    t.enableVertexAttribArray(F), t.bindBuffer(t.ARRAY_BUFFER, u), t.vertexAttribPointer(F, 3, t.FLOAT, !1, 0, 0), t.enableVertexAttribArray(g), t.bindBuffer(t.ARRAY_BUFFER, x), t.vertexAttribPointer(g, 3, t.FLOAT, !1, 0, 0), t.enableVertexAttribArray(S), t.bindBuffer(t.ARRAY_BUFFER, l), t.vertexAttribPointer(S, 2, t.FLOAT, !1, 0, 0), t.bindBuffer(t.ELEMENT_ARRAY_BUFFER, f);
    const B = t.getUniformLocation(c, "projectionMatrix");
    t.uniformMatrix4fv(B, !1, T.projectionMatrix.elements);
    const E = t.getUniformLocation(c, "u_plane_width");
    t.uniform1f(E, Y);
    const C = t.getUniformLocation(c, "u_plane_height");
    t.uniform1f(C, V);
    const U = t.getUniformLocation(c, "u_colors_count");
    t.uniform1i(U, $);
    const z = [
      "projectionMatrix",
      "modelViewMatrix",
      "u_time",
      "u_resolution",
      "u_color_pressure",
      "u_wave_frequency_x",
      "u_wave_frequency_y",
      "u_wave_amplitude",
      "u_colors_count",
      "u_plane_width",
      "u_plane_height",
      "u_shadows",
      "u_highlights",
      "u_grain_intensity",
      "u_grain_sparsity",
      "u_grain_scale",
      "u_grain_speed",
      "u_flow_distortion_a",
      "u_flow_distortion_b",
      "u_flow_scale",
      "u_flow_ease",
      "u_flow_enabled",
      "u_y_offset",
      "u_y_offset_wave_multiplier",
      "u_y_offset_color_multiplier",
      "u_y_offset_flow_multiplier",
      "u_procedural_texture",
      "u_enable_procedural_texture",
      "u_texture_ease",
      "u_transparent_texture_void",
      "u_saturation",
      "u_brightness",
      "u_color_blending",
      "u_domain_warp_enabled",
      "u_domain_warp_intensity",
      "u_domain_warp_scale",
      "u_vignette_intensity",
      "u_vignette_radius",
      "u_fresnel_enabled",
      "u_fresnel_power",
      "u_fresnel_intensity",
      "u_fresnel_color",
      "u_iridescence_enabled",
      "u_iridescence_intensity",
      "u_iridescence_speed",
      "u_bloom_intensity",
      "u_bloom_threshold",
      "u_chromatic_aberration",
      "u_shape_type",
      "u_silhouette_fade",
      "u_cylinder_fade",
      "u_ribbon_fade",
      "u_flat_shading"
    ], M = {
      attributes: { position: F, normal: g, uv: S },
      uniforms: {}
    };
    z.forEach((P) => {
      M.uniforms[P] = t.getUniformLocation(c, P);
    });
    for (let P = 0; P < $; P++)
      M.uniforms[`u_colors[${P}].is_active`] = t.getUniformLocation(c, `u_colors[${P}].is_active`), M.uniforms[`u_colors[${P}].color`] = t.getUniformLocation(c, `u_colors[${P}].color`), M.uniforms[`u_colors[${P}].influence`] = t.getUniformLocation(c, `u_colors[${P}].influence`);
    return this._initialized = !0, this._uniformsDirty = !0, this._colorsChanged = !0, this._textureDirty = !0, t.enable(t.BLEND), t.blendFunc(t.SRC_ALPHA, t.ONE_MINUS_SRC_ALPHA), t.enable(t.DEPTH_TEST), {
      gl: t,
      program: c,
      buffers: {
        position: u,
        normal: x,
        uv: l,
        index: f,
        wireframeIndex: R
      },
      locations: M,
      camera: T,
      indexCount: w.length,
      wireframeIndexCount: v.length,
      indexType: w instanceof Uint32Array ? t.UNSIGNED_INT : t.UNSIGNED_SHORT
    };
  }
  _createProceduralTexture(e) {
    this._sourceCanvas || (this._sourceCanvas = document.createElement("canvas"), this._sourceCanvas.width = 1024, this._sourceCanvas.height = 1024, this._sourceCtx = this._sourceCanvas.getContext("2d"));
    const s = this._sourceCanvas, r = this._sourceCtx;
    if (!r)
      return null;
    let t = this._textureSeed;
    const a = this._textureSeed;
    function o() {
      const g = Math.sin(t++) * 1e4;
      return g - Math.floor(g);
    }
    const _ = (g) => {
      t = a + g;
    }, m = this._colors.filter((g) => g.enabled).map((g) => g.color);
    if (m.length === 0)
      return null;
    const w = this._shapeType !== "plane", v = w ? [-1, 0, 1] : [0], u = w ? [-1, 0, 1] : [0];
    function x(g) {
      const S = parseInt(g.replace("#", ""), 16);
      return {
        r: S >> 16 & 255,
        g: S >> 8 & 255,
        b: S & 255
      };
    }
    function l(g, S, B) {
      return "#" + ((1 << 24) + (Math.round(g) << 16) + (Math.round(S) << 8) + Math.round(B)).toString(16).slice(1).padStart(6, "0");
    }
    const f = () => {
      const g = m[Math.floor(o() * m.length)], S = m[Math.floor(o() * m.length)], B = o() * this._textureColorBlending, E = x(g), C = x(S), U = E.r + (C.r - E.r) * B, z = E.g + (C.g - E.g) * B, M = E.b + (C.b - E.b) * B;
      return l(U, z, M);
    }, R = this._proceduralBackgroundColor || "#000000";
    r.fillStyle = R, r.fillRect(0, 0, 1024, 1024);
    const d = r.createLinearGradient(0, 0, 0, 1024);
    d.addColorStop(0, f()), d.addColorStop(1, f()), r.fillStyle = d, r.fillRect(0, 0, 1024, 1024);
    for (let g = 0; g < this._textureShapeTriangles; g++) {
      const S = f(), B = o() * 1024, E = o() * 1024, C = 100 + o() * 300, U = (o() - 0.5) * C, z = (o() - 0.5) * C, M = (o() - 0.5) * C, P = (o() - 0.5) * C;
      for (const I of v)
        for (const k of u) {
          r.fillStyle = S, r.beginPath();
          const D = B + I * 1024, W = E + k * 1024;
          r.moveTo(D, W), r.lineTo(D + U, W + z), r.lineTo(D + M, W + P), r.fill();
        }
    }
    for (let g = 0; g < this._textureShapeCircles; g++) {
      const S = f(), B = 10 + o() * 50, E = o() * 1024, C = o() * 1024, U = 50 + o() * 150;
      for (const z of v)
        for (const M of u)
          r.strokeStyle = S, r.lineWidth = B, r.beginPath(), r.arc(E + z * 1024, C + M * 1024, U, 0, Math.PI * 2), r.stroke();
    }
    for (let g = 0; g < this._textureShapeBars; g++) {
      const S = f(), B = o() * 1024, E = o() * 1024, C = o() * Math.PI;
      for (const U of v)
        for (const z of u)
          r.fillStyle = S, r.save(), r.translate(B + U * 1024, E + z * 1024), r.rotate(C), r.fillRect(-150, -25, 300, 50), r.restore();
    }
    r.lineWidth = 15, r.lineCap = "round";
    for (let g = 0; g < this._textureShapeSquiggles; g++) {
      const S = f(), B = o() * 1024, E = o() * 1024, C = [];
      let U = 0, z = 0;
      for (let M = 0; M < 4; M++) {
        const P = U + (o() - 0.5) * 300, I = z + (o() - 0.5) * 300;
        C.push({
          cx1: U + (o() - 0.5) * 300,
          cy1: z + (o() - 0.5) * 300,
          cx2: U + (o() - 0.5) * 300,
          cy2: z + (o() - 0.5) * 300,
          ex: P,
          ey: I
        }), U = P, z = I;
      }
      for (const M of v)
        for (const P of u) {
          r.strokeStyle = S, r.beginPath();
          const I = B + M * 1024, k = E + P * 1024;
          r.moveTo(I, k);
          for (const D of C)
            r.bezierCurveTo(
              I + D.cx1,
              k + D.cy1,
              I + D.cx2,
              k + D.cy2,
              I + D.ex,
              k + D.ey
            );
          r.stroke();
        }
    }
    _(5e4), this._maskedCanvas || (this._maskedCanvas = document.createElement("canvas"), this._maskedCanvas.width = 1024, this._maskedCanvas.height = 1024, this._maskedCtx = this._maskedCanvas.getContext("2d"));
    const b = this._maskedCanvas, A = this._maskedCtx;
    if (!A)
      return null;
    this._transparentTextureVoid ? A.clearRect(0, 0, 1024, 1024) : (A.fillStyle = R, A.fillRect(0, 0, 1024, 1024));
    let y = 0;
    const c = [];
    for (; y < 1024; )
      if (o() < this._textureVoidLikelihood) {
        const S = this._textureVoidWidthMin + o() * (this._textureVoidWidthMax - this._textureVoidWidthMin);
        c.push({ type: "void", x: y, width: S }), y += S;
      } else {
        const S = 50 + o() * 200;
        c.push({ type: "matter", x: y, width: S }), y += S;
      }
    for (const g of c)
      if (g.type === "matter") {
        const S = g.x, B = Math.min(g.x + g.width, 1024);
        let E = S;
        for (; E < B; ) {
          const C = (2 + o() * 20) / this._textureBandDensity, U = Math.floor(o() * 1024);
          A.drawImage(
            s,
            U,
            0,
            C,
            1024,
            E,
            0,
            C,
            1024
          ), E += C;
        }
      }
    const T = e.createTexture();
    e.bindTexture(e.TEXTURE_2D, T), e.texImage2D(e.TEXTURE_2D, 0, e.RGBA, e.RGBA, e.UNSIGNED_BYTE, b), e.texParameteri(e.TEXTURE_2D, e.TEXTURE_WRAP_S, e.REPEAT), e.texParameteri(e.TEXTURE_2D, e.TEXTURE_WRAP_T, e.REPEAT), e.texParameteri(e.TEXTURE_2D, e.TEXTURE_MIN_FILTER, e.LINEAR_MIPMAP_LINEAR), e.texParameteri(e.TEXTURE_2D, e.TEXTURE_MAG_FILTER, e.LINEAR), e.generateMipmap(e.TEXTURE_2D);
    const F = e.getExtension("EXT_texture_filter_anisotropic") || e.getExtension("MOZ_EXT_texture_filter_anisotropic") || e.getExtension("WEBKIT_EXT_texture_filter_anisotropic");
    if (F) {
      const g = e.getParameter(F.MAX_TEXTURE_MAX_ANISOTROPY_EXT);
      e.texParameterf(e.TEXTURE_2D, F.TEXTURE_MAX_ANISOTROPY_EXT, Math.min(16, g));
    }
    return T;
  }
  get fresnelColor() {
    return this._fresnelColor;
  }
  set fresnelColor(e) {
    this._fresnelColor !== e && (this._fresnelColor = e, this._fresnelColorRgb = this._hexToRgb(e), this._uniformsDirty = !0);
  }
  get shapeType() {
    return this._shapeType;
  }
  set shapeType(e) {
    this._shapeType !== e && (this._shapeType = e, this._updateGeometry());
  }
  get cameraLock() {
    return this._cameraLock;
  }
  set cameraLock(e) {
    this._cameraLock = e;
  }
  get cameraZoom() {
    return this._cameraZoom;
  }
  set cameraZoom(e) {
    this._cameraZoom !== e && (this._cameraZoom = e, this._updateCameraFrustum());
  }
  _updateCameraFrustum() {
    if (!this.glState)
      return;
    const e = this.glState.gl, i = this._ref.clientWidth, s = this._ref.clientHeight;
    j(this.glState.camera, i, s, Y, V, this._shapeType, this._cameraZoom);
    const r = this.glState.locations.uniforms.projectionMatrix;
    e.useProgram(this.glState.program), r && e.uniformMatrix4fv(r, !1, this.glState.camera.projectionMatrix.elements), this._uniformsDirty = !0;
  }
  _initWatermark() {
    const e = this.glState.gl, i = e, s = typeof i.createVertexArray == "function", r = e.createShader(e.VERTEX_SHADER);
    e.shaderSource(r, Tt), e.compileShader(r);
    const t = e.createShader(e.FRAGMENT_SHADER);
    e.shaderSource(t, Et), e.compileShader(t);
    const a = e.createProgram();
    e.attachShader(a, r), e.attachShader(a, t), e.linkProgram(a), this._watermarkProgram = a, e.deleteShader(r), e.deleteShader(t);
    const o = 13, _ = 6, m = 5, w = document.createElement("canvas").getContext("2d");
    w.font = `bold ${o}px -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif`;
    const v = w.measureText("NEAT"), u = Math.ceil(v.width), x = o, l = u + _ * 2, f = x + m * 2;
    this._watermarkWidth = l, this._watermarkHeight = f;
    const R = document.createElement("canvas");
    R.width = l, R.height = f;
    const d = R.getContext("2d");
    d.clearRect(0, 0, l, f), d.shadowColor = "rgba(0,0,0,0.4)", d.shadowBlur = 2, d.shadowOffsetX = 1, d.shadowOffsetY = 1, d.font = `bold ${o}px -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif`, d.textAlign = "center", d.textBaseline = "middle", d.fillStyle = "rgba(255,255,255,0.5)", d.fillText("NEAT", l / 2, f / 2);
    const b = e.createTexture();
    e.activeTexture(e.TEXTURE2), e.bindTexture(e.TEXTURE_2D, b), e.texParameteri(e.TEXTURE_2D, e.TEXTURE_WRAP_S, e.CLAMP_TO_EDGE), e.texParameteri(e.TEXTURE_2D, e.TEXTURE_WRAP_T, e.CLAMP_TO_EDGE), e.texParameteri(e.TEXTURE_2D, e.TEXTURE_MIN_FILTER, e.LINEAR), e.texParameteri(e.TEXTURE_2D, e.TEXTURE_MAG_FILTER, e.LINEAR), e.pixelStorei(e.UNPACK_PREMULTIPLY_ALPHA_WEBGL, !0), e.texImage2D(e.TEXTURE_2D, 0, e.RGBA, e.RGBA, e.UNSIGNED_BYTE, R), e.pixelStorei(e.UNPACK_PREMULTIPLY_ALPHA_WEBGL, !1), this._watermarkTexture = b;
    const A = e.createBuffer();
    e.bindBuffer(e.ARRAY_BUFFER, A), e.bufferData(e.ARRAY_BUFFER, new Float32Array([0, 1, 1, 1, 0, 0, 1, 0]), e.STATIC_DRAW), this._watermarkTexCoordBuffer = A;
    const y = e.createBuffer();
    if (e.bindBuffer(e.ARRAY_BUFFER, y), e.bufferData(e.ARRAY_BUFFER, new Float32Array(8), e.DYNAMIC_DRAW), this._watermarkBuffer = y, this._wmLocPos = e.getAttribLocation(a, "a_wm_position"), this._wmLocTc = e.getAttribLocation(a, "a_wm_texcoord"), this._wmLocTex = e.getUniformLocation(a, "u_wm_texture"), s) {
      this._watermarkVAO = i.createVertexArray(), i.bindVertexArray(this._watermarkVAO), e.enableVertexAttribArray(this._wmLocPos), e.bindBuffer(e.ARRAY_BUFFER, y), e.vertexAttribPointer(this._wmLocPos, 2, e.FLOAT, !1, 0, 0), e.enableVertexAttribArray(this._wmLocTc), e.bindBuffer(e.ARRAY_BUFFER, A), e.vertexAttribPointer(this._wmLocTc, 2, e.FLOAT, !1, 0, 0), this._gradientVAO = i.createVertexArray(), i.bindVertexArray(this._gradientVAO);
      const c = this.glState.locations.attributes;
      e.enableVertexAttribArray(c.position), e.bindBuffer(e.ARRAY_BUFFER, this.glState.buffers.position), e.vertexAttribPointer(c.position, 3, e.FLOAT, !1, 0, 0), e.enableVertexAttribArray(c.normal), e.bindBuffer(e.ARRAY_BUFFER, this.glState.buffers.normal), e.vertexAttribPointer(c.normal, 3, e.FLOAT, !1, 0, 0), e.enableVertexAttribArray(c.uv), e.bindBuffer(e.ARRAY_BUFFER, this.glState.buffers.uv), e.vertexAttribPointer(c.uv, 2, e.FLOAT, !1, 0, 0), e.bindBuffer(e.ELEMENT_ARRAY_BUFFER, this.glState.buffers.index), i.bindVertexArray(this._gradientVAO);
    }
    this._wmClickHandler = (c) => {
      this._isOverWatermark(c) && (c.preventDefault(), c.stopPropagation(), window.open("https://neat.firecms.co", "_blank", "noopener"));
    }, this._wmMoveHandler = (c) => {
      this._wmMoveRafPending || (this._wmMoveRafPending = !0, requestAnimationFrame(() => {
        this._wmMoveRafPending = !1;
        const T = performance.now();
        (!this._wmCachedRect || T - this._wmRectCacheTime > 100) && (this._wmCachedRect = this._ref.getBoundingClientRect(), this._wmRectCacheTime = T);
        const F = this._wmCachedRect, g = c.clientX - F.left, S = c.clientY - F.top, B = F.width, E = F.height;
        if (g < 0 || S < 0 || g > B || S > E) {
          this._ref.style.cursor === "pointer" && (this._ref.style.cursor = ""), document.body.style.cursor === "pointer" && (document.body.style.cursor = "");
          return;
        }
        const C = this._watermarkMargin, U = this._watermarkWidth, z = this._watermarkHeight, M = B - C - U, P = E - C - z, I = g >= M && g <= B - C && S >= P && S <= E - C;
        this._ref.style.cursor = I ? "pointer" : "", I ? document.body.style.cursor = "pointer" : document.body.style.cursor === "pointer" && (document.body.style.cursor = "");
      }));
    }, document.addEventListener("click", this._wmClickHandler, !0), document.addEventListener("mousemove", this._wmMoveHandler);
  }
  _isOverWatermark(e) {
    const i = performance.now();
    (!this._wmCachedRect || i - this._wmRectCacheTime > 100) && (this._wmCachedRect = this._ref.getBoundingClientRect(), this._wmRectCacheTime = i);
    const s = this._wmCachedRect, r = e.clientX - s.left, t = e.clientY - s.top, a = s.width, o = s.height;
    if (r < 0 || t < 0 || r > a || t > o)
      return !1;
    const _ = this._watermarkMargin, m = this._watermarkWidth, w = this._watermarkHeight, v = a - _ - m, u = o - _ - w;
    return r >= v && r <= a - _ && t >= u && t <= o - _;
  }
  _renderWatermark(e) {
    const i = this._watermarkProgram, s = this._watermarkTexture, r = this._watermarkBuffer;
    if (!i || !s || !r)
      return;
    const t = this._ref.width || this._ref.clientWidth, a = this._ref.height || this._ref.clientHeight;
    if (t === 0 || a === 0)
      return;
    const o = 4, _ = this._watermarkWidth, m = this._watermarkHeight, w = 1 - o / t * 2, v = w - _ / t * 2, u = -1 + o / a * 2, x = u + m / a * 2, l = this._wmPosData;
    l[0] = v, l[1] = u, l[2] = w, l[3] = u, l[4] = v, l[5] = x, l[6] = w, l[7] = x, e.bindBuffer(e.ARRAY_BUFFER, r), e.bufferSubData(e.ARRAY_BUFFER, 0, l);
    const f = e, R = this._watermarkVAO !== null;
    if (e.useProgram(i), e.disable(e.DEPTH_TEST), e.blendFunc(e.ONE, e.ONE_MINUS_SRC_ALPHA), R ? (f.bindVertexArray(this._watermarkVAO), e.bindBuffer(e.ARRAY_BUFFER, r), e.vertexAttribPointer(this._wmLocPos, 2, e.FLOAT, !1, 0, 0)) : (e.enableVertexAttribArray(this._wmLocPos), e.bindBuffer(e.ARRAY_BUFFER, r), e.vertexAttribPointer(this._wmLocPos, 2, e.FLOAT, !1, 0, 0), e.enableVertexAttribArray(this._wmLocTc), e.bindBuffer(e.ARRAY_BUFFER, this._watermarkTexCoordBuffer), e.vertexAttribPointer(this._wmLocTc, 2, e.FLOAT, !1, 0, 0)), e.activeTexture(e.TEXTURE2), e.bindTexture(e.TEXTURE_2D, s), e.uniform1i(this._wmLocTex, 2), e.drawArrays(e.TRIANGLE_STRIP, 0, 4), e.enable(e.DEPTH_TEST), e.blendFunc(e.SRC_ALPHA, e.ONE_MINUS_SRC_ALPHA), e.useProgram(this.glState.program), R)
      f.bindVertexArray(this._gradientVAO);
    else {
      const d = this.glState.locations.attributes;
      e.enableVertexAttribArray(d.position), e.bindBuffer(e.ARRAY_BUFFER, this.glState.buffers.position), e.vertexAttribPointer(d.position, 3, e.FLOAT, !1, 0, 0), e.enableVertexAttribArray(d.normal), e.bindBuffer(e.ARRAY_BUFFER, this.glState.buffers.normal), e.vertexAttribPointer(d.normal, 3, e.FLOAT, !1, 0, 0), e.enableVertexAttribArray(d.uv), e.bindBuffer(e.ARRAY_BUFFER, this.glState.buffers.uv), e.vertexAttribPointer(d.uv, 2, e.FLOAT, !1, 0, 0), e.bindBuffer(e.ELEMENT_ARRAY_BUFFER, this.glState.buffers.index);
    }
  }
}
for (const [h, e, i, s, r] of wt)
  Object.defineProperty(Rt.prototype, h, {
    get() {
      return i === 1 ? this[e] : this[e] * i;
    },
    set(t) {
      const a = s === 1 ? t : t * s;
      this[e] !== a && (this[e] = a, this._uniformsDirty = !0, r === "t" && this._enableProceduralTexture ? this._textureNeedsUpdate = !0 : r === "g" && this._updateGeometry());
    },
    enumerable: !0,
    configurable: !0
  });
function At() {
  const h = new Date(), e = h.getMinutes(), i = h.getSeconds();
  return e * 60 + i;
}
function St() {
  if (document.querySelector('meta[name="generator"][content*="NEAT"]'))
    return;
  const h = document.createElement("meta");
  h.name = "generator", h.content = "NEAT by FireCMS \u2014 https://neat.firecms.co", document.head.appendChild(h);
}
const Tt = `
attribute vec2 a_wm_position;
attribute vec2 a_wm_texcoord;
varying vec2 v_wm_texcoord;
void main() {
    gl_Position = vec4(a_wm_position, 0.0, 1.0);
    v_wm_texcoord = a_wm_texcoord;
}
`, Et = `
precision mediump float;
varying vec2 v_wm_texcoord;
uniform sampler2D u_wm_texture;
void main() {
    gl_FragColor = texture2D(u_wm_texture, v_wm_texcoord);
}
`;
export {
  Rt as NeatGradient
};
